client/mm/libxc/binance.go

```diff
@@ -0,0 +1,1168 @@
+		bnc.booksMtx.Lock()
+		_, exists := bnc.books[slug]
+		if !exists {
+			bnc.books[slug] = newBNCBook()
+		} else {
+			bnc.books[slug].numSubscribers++
+		}
+		bnc.booksMtx.Unlock()
+
+		if err := bnc.subUnsubDepth(bnc.marketStream, "SUBSCRIBE", slug); err != nil {
+			return fmt.Errorf("subUnsubDepth(SUBSCRIBE): %v", err)
+		}
+		return nil
+	}
+
+	streamHandler := func(b []byte) {
+		var note *bncBookNote
+		if err := json.Unmarshal(b, &note); err != nil {
+			bnc.log.Errorf("error unmarshaling book note: %v", err)
+			return
+		}
+
+		if note == nil || note.Data == nil {
+			bnc.log.Debugf("no data in %q update: %+v", slug, note)
+			return
+		}
+
+		parts := strings.Split(note.StreamName, "@")
+		if len(parts) != 2 || parts[1] != "depth20" {
+			bnc.log.Errorf("unknown stream name %q", note.StreamName)
+			return
+		}
+		slug = parts[0] // will be lower-case
+
+		bnc.log.Tracef("received %d bids and %d asks in a book update for %s", len(note.Data.Bids), len(note.Data.Asks), slug)
+
+		bids, err := bncParseBookUpdates(note.Data.Bids)
+		if err != nil {
+			bnc.log.Errorf("error parsing bid updates: %v", err)
+			return
+		}
+
+		asks, err := bncParseBookUpdates(note.Data.Asks)
+		if err != nil {
+			bnc.log.Errorf("error parsing ask updates: %v", err)
+			return
+		}
+
+		bnc.booksMtx.Lock()
+		defer bnc.booksMtx.Unlock()
+		book := bnc.books[slug]
+		if book == nil {
+			bnc.log.Errorf("no book for stream %q", slug)
+			return
+		}
+
+		book.asks = asks
+		book.bids = bids
+	}
+
+	newConn := func() (comms.WsConn, *dex.ConnectionMaster, error) {
+		if bnc.ctx.Err() != nil {
+			return nil, nil, bnc.ctx.Err()
+		}
+		bnc.log.Debugf("Creating a new Binance market stream handler")
+		// Get a list of current subscriptions so we can restart them all.
+		bnc.booksMtx.Lock()
+		bnc.books[slug] = newBNCBook()
+		bnc.booksMtx.Unlock()
+
+		streamName := slug + "@depth20"
+		addr := fmt.Sprintf("%s/stream?streams=%s", bnc.wsURL, streamName)
+
+		// Need to send key but not signature
+		conn, err := comms.NewWsConn(&comms.WsCfg{
+			URL: addr,
+			// The websocket server will send a ping frame every 3 minutes. If the
+			// websocket server does not receive a pong frame back from the
+			// connection within a 10 minute period, the connection will be
+			// disconnected. Unsolicited pong frames are allowed.
+			PingWait: time.Minute * 4,
+			ReconnectSync: func() {
+				bnc.log.Debugf("Binance reconnected")
+			},
+			ConnectEventFunc: func(cs comms.ConnectionStatus) {},
+			Logger:           bnc.log.SubLogger("BNCBOOK"),
+			RawHandler:       streamHandler,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cm := dex.NewConnectionMaster(conn)
+		if err = cm.ConnectOnce(bnc.ctx); err != nil {
+			return nil, nil, fmt.Errorf("websocketHandler remote connect: %v", err)
+		}
+
+		return conn, cm, nil
+	}
+
+	conn, cm, err := newConn()
+	if err != nil {
+		return fmt.Errorf("error initializing connection: %v", err)
+	}
+	bnc.marketStream = conn
```

Comment:
On this one I definitely agree, it should start a retry loop.

Reply:
Actually I have the same opinion for this one as well.. if on the initial connect it does not work, it should fail loudly, so that the user can fix it. If the connection goes down later though when likely the user is not around, it should keep trying to reconnect.